Core/include/Acts/Detector/ComponentBuilderProxy.hpp

```diff
@@ -0,0 +1,168 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Geometry/VolumeBounds.hpp"
+#include "Acts/Utilities/BinningData.hpp"
+
+#include <memory>
+#include <string>
+
+namespace Acts {
+namespace Experimental {
+
+class IDetectorComponentBuilder;
+
+/// @brief A container proxy struct that can be used to build up
+/// a tree graph of detector components
+class ComponentBuilderProxy
+    : public std::enable_shared_from_this<ComponentBuilderProxy> {
+ protected:
+  /// Createing a ComponentBuilderProxy is not allowed
+  /// @note use : createRootProxy(), createChildVolumeProxy() or
+  /// createChildContainerProxy()
+  ComponentBuilderProxy() = default;
+
+ public:
+  /// @brief Private constructor to be used for the proxy creation
+  /// @param n The name of the proxy
+  ComponentBuilderProxy(const std::string& n) : name(n){};
+
+  /// @brief  Definition of a volume
+  struct VolumeProxy {
+    /// If fully defined, the builder is set
+    std::shared_ptr<const IDetectorComponentBuilder> builder = nullptr;
+  };
+
+  /// @brief  Definition of a container
+  struct ContainerProxy {
+    /// The binning instruction
+    std::vector<BinningValue> binning = {};
+    /// The contained component builders, in case of a container builder
+    std::vector<std::shared_ptr<ComponentBuilderProxy>> children = {};
+  };
+
+  /// The proxy name - this needs to be uniuque
+  std::string name = "";
```

Comment:
Do these have to be settable from the outside? Otherwise I'd suggest following the usual naming of `m_name` etc, and then also renaming the method arguments in this class to not be single letters, where possible.

Reply:
In general, if the name of the proxies are used to then generate the builder tools (that's their only reason to be there), then it could be useful to harmonise namings when the full tree is constructed, but again, that's not really necessary.

Reply:
I see your point. In the rest of the code base we really only expose members if we expect them to be set from other classes etc, and we prefix internal members with `m_`, among other things so we have fewer naming issues, and that it's clear that it's a member variable.